### PR TITLE
SEAB-3697: update topicAutomatic on refresh

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
UPDATE: This PR has changed per review feedback.  New description follows:

This PR modifies the webservice to update topicAutomatic on refresh for workflows, and on creation and refresh for tools.   TopicManual and topicSelection are not changed.

For workflows, the implementation is a one-liner which adds a copy of topicAutomatic in Entry.update, which is called by the refresh code to copy information from a temporary workflow object to the result workflow object.

For tools, I added code to harvest/set the topic after the tools are saved to the db, at various points in both the refresh and registration code.  In the interest of clarity/elegance, my original goal was a single point of modification, but I searched for a while, and I'm not sure there is one.  In general, the existing tools code might benefit from a refactor.

NOTE TO REVIEWER: I have measured a delay of up to 20 seconds between when you submit a new "About" description on the GitHub repo page, and when that new description propagates to the GitHub API so it registers during a workflow refresh.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-3697

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] Check the Snyk dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

### CODE REVIEW

File: dockstore-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java [hunk 1]
```diff
@@ -644,6 +644,41 @@ public void testAddingWorkflowForumUrlAndTopic() throws ApiException {
         Assert.assertEquals(TopicSelectionEnum.MANUAL, updatedWorkflow.getTopicSelection());
     }
 
+    @Test
+    public void testTopicAfterRefresh() throws ApiException {
+        ApiClient webClient = WorkflowIT.getWebClient(USER_2_USERNAME, testingPostgres);
+        WorkflowsApi workflowsApi = new WorkflowsApi(webClient);
+
+        Workflow workflow = manualRegisterAndPublish(workflowsApi, "DockstoreTestUser2/hello-dockstore-workflow", "testname", "cwl",
```
Comment (garyluu) [inline dockstore-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java:652]: There's an enum for "cwl" or "CWL" somewhere

**Review by garyluu — APPROVED** [review]

**Review by denis-yuen — COMMENTED** [review]
```
I wrote an integration test for workflows, but not one for tools, since the existing code doesn't appear to harvest topicAutomatic from anywhere upon tool creation, at present.
```

Since "github apps but for tools" is still in progress, I suspect that we aren't harvesting topics upon tool creation (i.e. github app install) 

That said, will this change pull in topics on refresh for existing (non-github app tools)?

**Comment by svonworl** [comment]
> ```
> I wrote an integration test for workflows, but not one for tools, since the existing code doesn't appear to harvest topicAutomatic from anywhere upon tool creation, at present.
> ```
> 
> Since "github apps but for tools" is still in progress, I suspect that we aren't harvesting topics upon tool creation (i.e. github app install)
> 
> That said, will this change pull in topics on refresh for existing (non-github app tools)?

No, it doesn't pull in topics for existing tools, for the following reason:

The current `refresh` code works by creating a temporary Entry object T, the same way it would create a new Entry, and then copies the attributes of T to the existing Entry using `update()`.  So, if the topic is harvested during Entry creation, it will be updated on refresh, too.

Since the existing code doesn't harvest the topic for new tools, the information isn't there to be copied by `update`.  I could do a hack to retrieve the tool topic only on refresh, but I believe the correct approach is to implement the topic harvest upon tool creation, in which case, refresh copy the topic during the `update()`, no other mods required.

**Comment by denis-yuen** [comment]
> Since the existing code doesn't harvest the topic for new tools, the information isn't there to be copied by `update`. I could do a hack to retrieve the tool topic only on refresh, but I believe the correct approach is to implement the topic harvest upon tool creation, in which case, a refresh will copy the topic during the `update()`, no other mods required.

Fair enough and makes sense from a code perspective (i.e. creation separate from update). 
From the perspective of the user though, I think the intent of the ticket was to get this working when someone hits the refresh (regardless of tool or workflow) button, so let's get that in to finish the ticket (up to you whether it should get an independent PR)

**Review by coverbeck — COMMENTED** [review]
Looks good so far, but just saw Denis' comments, so I'm going to wait for that. Please re-tag me after you do that (or if you decide to do it separately), and I'll go over it again.

**Comment by svonworl** [comment]
> Looks good so far, but just saw Denis' comments, so I'm going to wait for that. Please re-tag me after you do that (or if you decide to do it separately), and I'll go over it again.

This PR is updated to set the automatic topic for both workflows and tools.

**Comment by svonworl** [comment]
ugh, the commented-out liquibase pom line propagated to the PR, fixing...

**Review by coverbeck — APPROVED** [review]

**Review by garyluu — APPROVED** [review]

**Review by kathy-t — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]

**Pull request merged**